src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -461,18 +461,95 @@ private void ProcessRenderQueue()
             }
         }
 
-        private Task InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents)
+        private Task InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents, Task updateDisplayTask)
         {
+            if (updateDisplayTask.IsCanceled)
+            {
+                // The display update was cancelled (maybe due to a timeout on the components server-side case or due
+                // to the renderer being disposed)
+                return Task.CompletedTask;
+            }
+            if (updateDisplayTask.IsFaulted)
+            {
+                // The display update failed so we don't care any more about running on render completed
+                // fallbacks as the entire rendering process is going to be torn down.
+                HandleException(updateDisplayTask.Exception);
+                return Task.CompletedTask;
+            }
+
+            if (!updateDisplayTask.IsCompleted)
+            {
+                var updatedComponentsId = new int[updatedComponents.Count];
+                var updatedComponentsArray = updatedComponents.Array;
+                for (int i = 0; i < updatedComponentsId.Length; i++)
+                {
+                    updatedComponentsId[i] = updatedComponentsArray[i].ComponentId;
+                }
+
+                return InvokeAfterUpdateDisplayCompleted(updateDisplayTask, updatedComponentsId);
+            }
+
             List<Task> batch = null;
             var array = updatedComponents.Array;
             for (var i = 0; i < updatedComponents.Count; i++)
             {
                 var componentState = GetOptionalComponentState(array[i].ComponentId);
                 if (componentState != null)
+                {
+                    ProcessComponent(componentState, ref batch);
+                }
+            }
+
+            return batch != null ?
+                Task.WhenAll(batch) :
+                Task.CompletedTask;
+
+            async Task InvokeAfterUpdateDisplayCompleted(Task updateDisplayTask, int[] updatedComponents)
```

Comment:
Would it be OK for `InvokeAfterUpdateDisplayCompleted` and `ProcessComponent` *not* to be local functions, but rather to be regular class-level methods?

Local functions are always harder to reason about because you have to read every single line of code carefully to figure out whether it's hoisting in some local var (which if it does, changes the execution characteristics significantly, e.g., forcing an allocation at the call site).

Also then it would be handy to give them more descriptive names. E.g., `InvokeAfterUpdateDisplayCompleted` could be `InvokeRenderCompletedCallsAfterTask`, and `ProcessComponent` could be `NotifyRenderCompletedAsync`.

Reply:
Id personally rather use local functions because these methods only make sense in the context of the enclosing method, The hoisting can easily be solved by marking them as static (C# 8). That way you don't have three methods in the class scope adding noise.

I usually take special care of not hoisting anything in local functions unless there is an advantage to it. I do believe though that local functions are converted to structs, so I'm not sure the allocations apply.

That said, I don't feel very strongly about this, so I'll just change it.

Reply:
done